From: sunshineinabox <aqemail@gmail.com>
Date: Mon, 27 Jul 2026 20:43:51 -0700
Subject: [PATCH] ASoC: codecs: aw88166: make volume control usable

- Invert the value to match userspace expectations
- Provide TLV metadata for the dB scale
- Set the default volume to the halfway point instead of full blast
- Simplify aw_dev_set_volume using regmap_update_bits
- Do not add the initial volume from the profile to the requested
  volume as that would throw off the dB mapping

Unlike the aw88261, the aw88166 volume field is
a plain linear attenuation counter in 0.125dB units,
so the full 0..1023 range maps to 0..-127.875dB.

With this change, it's actually possible to use this hardware volume
control as PlaybackVolume in an ALSA UCM profile.

Adapted for aw88166 from Val Packett's aw88261 series.
---
 sound/soc/codecs/aw88166.c | 43 +++++++++++++++++---------------------
 sound/soc/codecs/aw88166.h |  4 ++--
 2 files changed, 21 insertions(+), 26 deletions(-)
--- a/sound/soc/codecs/aw88166.c
+++ b/sound/soc/codecs/aw88166.c
@@ -15,6 +15,7 @@
 #include <linux/regmap.h>
 #include <sound/soc.h>
 #include <sound/pcm_params.h>
+#include <sound/tlv.h>
 #include "aw88166.h"
 #include "aw88395/aw88395_device.h"
 
@@ -613,24 +614,10 @@
 
 static int aw_dev_set_volume(struct aw_device *aw_dev, unsigned int value)
 {
-	struct aw_volume_desc *vol_desc = &aw_dev->volume_desc;
-	unsigned int reg_value;
-	u16 real_value;
-	int ret;
-
-	real_value = min((value + vol_desc->init_volume), (unsigned int)AW88166_MUTE_VOL);
-
-	ret = regmap_read(aw_dev->regmap, AW88166_SYSCTRL2_REG, &reg_value);
-	if (ret)
-		return ret;
-
-	dev_dbg(aw_dev->dev, "value 0x%x , reg:0x%x", value, real_value);
+	unsigned int volume = min(value, (unsigned int)AW88166_MUTE_VOL);
 
-	real_value = (real_value << AW88166_VOL_START_BIT) | (reg_value & AW88166_VOL_MASK);
-
-	ret = regmap_write(aw_dev->regmap, AW88166_SYSCTRL2_REG, real_value);
-
-	return ret;
+	return regmap_update_bits(aw_dev->regmap, AW88166_SYSCTRL2_REG,
+			~AW88166_VOL_MASK, volume << AW88166_VOL_START_BIT);
 }
 
 static void aw88166_dev_mute(struct aw_device *aw_dev, bool is_mute)
@@ -1434,7 +1421,8 @@
 	struct aw88166 *aw88166 = snd_soc_component_get_drvdata(codec);
 	struct aw_volume_desc *vol_desc = &aw88166->aw_pa->volume_desc;
 
-	ucontrol->value.integer.value[0] = vol_desc->ctl_volume;
+	ucontrol->value.integer.value[0] =
+		(AW88166_MUTE_VOL - vol_desc->ctl_volume) / 2;
 
 	return 0;
 }
@@ -1447,12 +1435,13 @@
 	struct aw_volume_desc *vol_desc = &aw88166->aw_pa->volume_desc;
 	struct soc_mixer_control *mc =
 		(struct soc_mixer_control *)kcontrol->private_value;
-	int value;
+	int value = ucontrol->value.integer.value[0];
 
-	value = ucontrol->value.integer.value[0];
 	if (value < mc->min || value > mc->max)
 		return -EINVAL;
 
+	value = AW88166_MUTE_VOL - (value * 2);
+
 	if (vol_desc->ctl_volume != value) {
 		vol_desc->ctl_volume = value;
 		aw_dev_set_volume(aw88166->aw_pa, vol_desc->ctl_volume);
@@ -1579,10 +1568,16 @@
 	return ret;
 }
 
+/*
+ * The volume register is in 0.125dB units which is too precise for
+ * TLV (!) so we have to multiply the scale by 2.
+ */
+static const DECLARE_TLV_DB_SCALE(volume_tlv, -12775, 25, 0);
+
 static const struct snd_kcontrol_new aw88166_controls[] = {
-	SOC_SINGLE_EXT("PCM Playback Volume", AW88166_SYSCTRL2_REG,
-		6, AW88166_MUTE_VOL, 0, aw88166_volume_get,
-		aw88166_volume_set),
+	SOC_SINGLE_EXT_TLV("PCM Playback Volume", AW88166_SYSCTRL2_REG,
+		0, AW88166_CTL_MAX_VOL, 1,
+		aw88166_volume_get, aw88166_volume_set, volume_tlv),
 	SOC_SINGLE_EXT("Calib", 0, 0, AW88166_CALI_RE_MAX, 0,
 		aw88166_re_get, aw88166_re_set),
 	AW88166_PROFILE_EXT("AW88166 Profile Set", aw88166_profile_info,
@@ -1701,7 +1696,7 @@
 	aw_dev->channel = AW88166_DEV_DEFAULT_CH;
 	aw_dev->fw_status = AW88166_DEV_FW_FAILED;
 
-	aw_dev->volume_desc.ctl_volume = AW88166_VOL_DEFAULT_VALUE;
+	aw_dev->volume_desc.ctl_volume = AW88166_CTL_DEFAULT_VOL;
 
 	aw88166_parse_channel_dt(aw88166);
 
--- a/sound/soc/codecs/aw88166.h
+++ b/sound/soc/codecs/aw88166.h
@@ -539,8 +539,8 @@
 #define AW88166_DSP_RE_SHIFT			(12)
 #define AW88166_CALI_RE_MAX			(15000)
 #define AW88166_CALI_RE_MIN			(4000)
-#define AW88166_VOLUME_STEP_DB			(64)
-#define AW88166_VOL_DEFAULT_VALUE		(0)
+#define AW88166_CTL_MAX_VOL		(AW88166_MUTE_VOL / 2)
+#define AW88166_CTL_DEFAULT_VOL		(AW88166_CTL_MAX_VOL / 2)
 #define AW88166_DSP_RE_TO_SHOW_RE(re, shift)	(((re) * (1000)) >> (shift))
 #define AW88166_SHOW_RE_TO_DSP_RE(re, shift)	(((re) << shift) / (1000))
 
